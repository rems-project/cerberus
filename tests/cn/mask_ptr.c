
/* A test case about modulus and masking, in both integers and pointers.
   Originally introduced as a test case because the pointer variant was causing
   issues for the SMT solver for unclear reasons.
*/

typedef unsigned long long u64;

enum {
  SHIFT_AMOUNT = 5
};

u64
foo_integer (u64 y)
/*@ requires mod(y, shift_left(1u64, ((u64) SHIFT_AMOUNT))) == 0u64; @*/
/* y = 42 */
/* shift_left(1u64, 5u64) = 0...100000*/
{
  u64 x = y;
  x &= ~ ((1UL << SHIFT_AMOUNT) - 1);
  /*@ assert (x == y); @*/
  return x;
}



int *
foo (int *p)
/*@ requires let p_u64 = (u64) p;
             mod(p_u64, shift_left(1u64, ((u64) SHIFT_AMOUNT))) == 0u64; @*/
{
  u64 x = ((u64) p);
  int *p2;

  x &= ~ ((1UL << SHIFT_AMOUNT) - 1);

  p2 = ((int *) x);
  /*@ assert (((u64) p2) == ((u64) p)); @*/
  return p2;
}

int main(void)
/*@ trusted; @*/
{
  int r1 = foo_integer(128);
  int p[1] = {512};
  int *r2 = foo(p);
}
