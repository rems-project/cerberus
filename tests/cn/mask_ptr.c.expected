return code: 0
[1/2]: foo_integer -- pass
[2/2]: foo -- pass
